lib/segment/src/vector_storage/multi_dense/appendable_mmap_multi_dense_vector_storage.rs

```diff
@@ -0,0 +1,271 @@
+            self.deleted_count -= 1;
+        }
+        Ok(previous)
+    }
+}
+
+impl<T: PrimitiveVectorElement> MultiVectorStorage<T> for AppendableMmapMultiDenseVectorStorage<T> {
+    fn get_multi(&self, key: PointOffsetType) -> TypedMultiDenseVectorRef<T> {
+        let mmap_offset = self.offsets.get(key as usize).unwrap().first().unwrap();
+        let inner_vector = self
+            .vectors
+            .get_many(mmap_offset.offset, mmap_offset.count as usize)
+            .expect("vector not found");
+        TypedMultiDenseVectorRef {
+            inner_vector,
+            dim: self.vectors.dim(),
+        }
+    }
+
+    fn multi_vector_config(&self) -> &MultiVectorConfig {
+        &self.multi_vector_config
+    }
+}
+
+impl<T: PrimitiveVectorElement> VectorStorage for AppendableMmapMultiDenseVectorStorage<T> {
+    fn vector_dim(&self) -> usize {
+        self.vectors.dim()
+    }
+
+    fn distance(&self) -> Distance {
+        self.distance
+    }
+
+    fn datatype(&self) -> VectorStorageDatatype {
+        T::datatype()
+    }
+
+    fn is_on_disk(&self) -> bool {
+        true
+    }
+
+    fn total_vector_count(&self) -> usize {
+        self.offsets.len()
+    }
+
+    fn get_vector(&self, key: PointOffsetType) -> CowVector {
+        // TODO(colbert) borrow instead of clone
+        let multivector = self.get_multi(key);
+        let multivector = TypedMultiDenseVector {
+            inner_vector: multivector.inner_vector.to_vec(),
+            dim: multivector.dim,
+        };
+        CowVector::MultiDense(T::into_float_multivector(Cow::Owned(multivector)))
+    }
+
+    fn insert_vector(&mut self, key: PointOffsetType, vector: VectorRef) -> OperationResult<()> {
+        let multi_vector: &MultiDenseVector = vector.try_into()?;
+        let multi_vector = T::from_float_multivector(Cow::Borrowed(multi_vector));
+        let multi_vector = multi_vector.as_ref();
+        assert_eq!(multi_vector.dim, self.vectors.dim());
+
+        let mut offset = self
+            .offsets
+            .get(key as usize)
+            .map(|x| x.first().copied().unwrap_or_default())
+            .unwrap_or_default();
+
+        if multi_vector.len() > offset.reserved as usize {
+            // append vector to the end
+            let mut new_key = self.vectors.len();
+            let chunk_left_keys = self.vectors.get_chunk_left_keys(new_key);
+            if multi_vector.len() > chunk_left_keys {
+                new_key += chunk_left_keys;
+            }
+
+            offset = MultivectorMmapOffset {
+                offset: new_key as PointOffsetType,
+                count: multi_vector.len() as PointOffsetType,
+                reserved: multi_vector.len() as PointOffsetType,
+            };
+        } else {
+            // use existing place to insert vector
+            offset.count = multi_vector.len() as PointOffsetType;
+        }
+
+        self.offsets.insert(key as usize, &[offset])?;
```

Comment:
what happened if we crash between updating `offsets` and `vectors`

Reply:
Corrupted storage will happend. I changed the order of insertion. I insert into `vectors` first and then into `offsets`. Thanks for this finding!